File: DialedInCoffee/RulesEngine/RulesEngine.swift
import Foundation
import OSLog

public struct RulesEngine {
    private static let log = Logger(subsystem: "coffee.dialedin", category: "RulesEngine")

    /// Secondary cause must reach this confidence and target a different parameter to be emitted.
    public static let secondaryConfidenceThreshold: Double = 0.6

    public static func evaluate(
        symptoms: [Symptom],
        current: MachineSettings,
        milk: Milk,
        drink: DrinkType,
        bean: BeanSnapshot? = nil,
        recentHistory: [HistorySnapshot] = [],
        novice: Bool = true
    ) -> Recommendation {
        log.debug("evaluate symptoms=\(symptoms.map(\.rawValue).joined(separator: ","), privacy: .public) drink=\(drink.rawValue, privacy: .public)")

        // 0. Empty input — nothing to do.
        guard !symptoms.isEmpty else {
            return Recommendation(rationale: "Pick at least one symptom to get a recommendation.")
        }

        // 1. Detect repeated failure → propose revert to last good.
        if LearningHeuristics.detectRepeatedFailure(history: recentHistory) {
            let lg = LearningHeuristics.lastGood(history: recentHistory)
            let rationale: String = {
                if lg != nil {
                    return "You've changed the same setting in the same direction several times without success. Consider going back to your last good recipe and changing something else."
                } else {
                    return "You've been pushing the same setting in one direction. Try the opposite direction or change a different setting instead."
                }
            }()
            log.debug("repeated failure detected; suggesting revert")
            return Recommendation(
                primary: nil,
                secondary: nil,
                topCause: nil,
                confidence: 0.8,
                rationale: rationale,
                contributions: [],
                alternativeCauses: [],
                suggestRevertToLastGood: true
            )
        }

        // 2. Aggregate causes from rules.
        let aggregated = CauseAggregator.aggregate(
            symptoms: symptoms,
            current: current,
            drink: drink,
            milk: milk,
            bean: bean
        )
        guard let top = aggregated.first else {
            return Recommendation(rationale: "No matching rules. Try describing the cup differently.")
        }

        // 3. Plan primary adjustment (with Australian-style override for milk drinks).
        var primary: Adjustment? = AdjustmentPlanner.australianBiasOverride(
            current: current,
            drink: drink,
            symptoms: symptoms
        )

        if primary == nil {
            primary = AdjustmentPlanner.adjustment(
                for: top.cause,
                current: current,
                drink: drink,
                milk: milk
            )
        }

        // 4. Plan optional secondary — must be a different parameter and clear above threshold.
        var secondary: Adjustment? = nil
        if let primary,
           let runnerUp = aggregated.dropFirst().first(where: { $0.confidence >= secondaryConfidenceThreshold }),
           let candidate = AdjustmentPlanner.adjustment(
                for: runnerUp.cause,
                current: current,
                drink: drink,
                milk: milk
           ),
           AdjustmentPlanner.differentParameter(candidate, from: primary)
        {
            secondary = candidate
        }

        // 5. Build contributions for Expert mode.
        let contributions = aggregated.map {
            CauseContribution(cause: $0.cause, confidence: $0.confidence, ruleIDs: $0.contributingRuleIDs)
        }
        let alternatives = Array(contributions.dropFirst().prefix(3))

        let rationale = ExplanationBuilder.rationale(
            topCause: top.cause,
            symptoms: symptoms,
            novice: novice
        )

        return Recommendation(
            primary: primary,
            secondary: secondary,
            topCause: top.cause,
            confidence: top.confidence,
            rationale: rationale,
            contributions: contributions,
            alternativeCauses: alternatives,
            suggestRevertToLastGood: false
        )
    }
}

File: android/core/src/main/kotlin/com/cawfee/domain/rules/RulesEngine.kt
package com.cawfee.domain.rules

import com.cawfee.domain.model.Adjustment
import com.cawfee.domain.model.BeanSnapshot
import com.cawfee.domain.model.CauseContribution
import com.cawfee.domain.model.DrinkType
import com.cawfee.domain.model.HistorySnapshot
import com.cawfee.domain.model.MachineSettings
import com.cawfee.domain.model.Milk
import com.cawfee.domain.model.Recommendation
import com.cawfee.domain.model.Symptom

/** The orchestrating rules engine. Ported 1:1 from RulesEngine.swift. */
object RulesEngine {

    /** Secondary cause must reach this confidence and target a different parameter. */
    const val SECONDARY_CONFIDENCE_THRESHOLD = 0.6

    fun evaluate(
        symptoms: List<Symptom>,
        current: MachineSettings,
        milk: Milk,
        drink: DrinkType,
        bean: BeanSnapshot? = null,
        recentHistory: List<HistorySnapshot> = emptyList(),
        novice: Boolean = true,
    ): Recommendation {
        // 0. Empty input.
        if (symptoms.isEmpty()) {
            return Recommendation(rationale = "Pick at least one symptom to get a recommendation.")
        }

        // 1. Repeated failure → propose revert to last good.
        if (LearningHeuristics.detectRepeatedFailure(recentHistory)) {
            val lg = LearningHeuristics.lastGood(recentHistory)
            val rationale = if (lg != null) {
                "You've changed the same setting in the same direction several times without success. " +
                    "Consider going back to your last good recipe and changing something else."
            } else {
                "You've been pushing the same setting in one direction. " +
                    "Try the opposite direction or change a different setting instead."
            }
            return Recommendation(
                confidence = 0.8,
                rationale = rationale,
                suggestRevertToLastGood = true,
            )
        }

        // 2. Aggregate causes.
        val aggregated = CauseAggregator.aggregate(symptoms, current, drink, milk, bean)
        val top = aggregated.firstOrNull()
            ?: return Recommendation(rationale = "No matching rules. Try describing the cup differently.")

        // 3. Plan primary (Australian-style override first).
        var primary: Adjustment? = AdjustmentPlanner.australianBiasOverride(current, drink, symptoms)
        if (primary == null) {
            primary = AdjustmentPlanner.adjustment(top.cause, current, drink, milk)
        }

        // 4. Optional secondary — different parameter, above threshold.
        var secondary: Adjustment? = null
        if (primary != null) {
            val runnerUp = aggregated.drop(1).firstOrNull { it.confidence >= SECONDARY_CONFIDENCE_THRESHOLD }
            if (runnerUp != null) {
                val candidate = AdjustmentPlanner.adjustment(runnerUp.cause, current, drink, milk)
                if (candidate != null && AdjustmentPlanner.differentParameter(candidate, primary)) {
                    secondary = candidate
                }
            }
        }

        // 5. Contributions for Expert mode.
        val contributions = aggregated.map {
            CauseContribution(cause = it.cause, confidence = it.confidence, ruleIDs = it.contributingRuleIDs)
        }
        val alternatives = contributions.drop(1).take(3)

        val rationale = ExplanationBuilder.rationale(top.cause, symptoms, novice)

        return Recommendation(
            primary = primary,
            secondary = secondary,
            topCause = top.cause,
            confidence = top.confidence,
            rationale = rationale,
            contributions = contributions,
            alternativeCauses = alternatives,
            suggestRevertToLastGood = false,
        )
    }
}
